### PR TITLE
Fixes #32631: Add keystore_certificate provider type

### PR DESCRIPTION
**Pull request by ehelms**
<del>Requires https://github.com/theforeman/puppet-certs/pull/334</del>

### CODE REVIEW

File: lib/puppet/provider/keystore/keytool.rb
Comment (ekohl) [inline lib/puppet/provider/keystore/keytool.rb:31]: Does error halt execution? It looks like it'll then proceed with attempting to delete the entry, even if it failed.

Comment (ehelms) [reply lib/puppet/provider/keystore/keytool.rb:31]: Addressed in the origin PR -- https://github.com/theforeman/puppet-certs/pull/334/files#diff-ab1ff29ab535e60c2e133829981732e7bdc416e5eabf07703b5b32e0ecafda51R18

File: manifests/candlepin.pp
Comment (evgeni) [inline manifests/candlepin.pp:140]: I must admit I find the notion of `$keystore:$alias` for the name rather confusing. Is this a common notation in the Java/keytool universe?

I see you could also write
```
keystore_certificate { "funny_name":
  keystore => $keystore,
  alias     => 'tomcat'
  …
}
```

And I find this more readable. But wouldn't block this on my weird brain :)

File: manifests/candlepin.pp
Comment (evgeni) [inline manifests/candlepin.pp:126]: unrelated to this PR (as this is old code), but does this leak the password to the logs?

File: lib/puppet/type/keystore_certificate.rb
Comment (evgeni) [inline lib/puppet/type/keystore_certificate.rb:36]:
```suggestion
    desc "Path to file containing the private key"
```

File: lib/puppet/type/keystore_certificate.rb
Comment (evgeni) [inline lib/puppet/type/keystore_certificate.rb:41]:
```suggestion
    desc "Path to the file containing th CA certificate"
```

(to be consistent with the other param docs)

File: manifests/candlepin.pp
Comment (evgeni) [inline manifests/candlepin.pp:145]: should *all* these be called `*_file`, as they are all… files?

File: manifests/candlepin.pp
Comment (ekohl) [reply manifests/candlepin.pp:140]: It's more of a Puppet convention. I think that what you write will also work with the current code. The benefit of the current version is that you can write `tomcat` into multiple keystores without having duplicate resources, so this works:
```puppet
keystore_certificate { "${keystore1}:tomcat":
}
keystore_certificate { "${keystore2}:tomcat":
}
```

If you don't, you need to use things like `funny_name1` and `funny_name2`. There are arguments for both.

That said, it's probably good to write a test case for both syntaxes to verify it really does work for both versions.

Comment (evgeni) [reply manifests/candlepin.pp:140]: Yeah, I also read the code that both notations *should* work.

File: manifests/candlepin.pp
Comment (ehelms) [reply manifests/candlepin.pp:126]: Solid question -- https://github.com/theforeman/puppet-certs/pull/337

File: manifests/candlepin.pp
Comment (ehelms) [reply manifests/candlepin.pp:145]: I have seen in other modules related to certs not attaching `_file` so I lean towards going with that model and it feels tidy.

**Comment by ehelms** [comment]
Rebased and updated based on comments

File: lib/puppet/provider/keystore_certificate/openssl.rb
Comment (ekohl) [inline lib/puppet/provider/keystore_certificate/openssl.rb:14]: I don't think this is a valid thing. I'd expect that `keystore_content` always returns a string which evaluates to true. Or does it rely on it raising a `Puppet::ExecutionFailure`if the alias isn't present?

File: lib/puppet/type/keystore_certificate.rb
Comment (ekohl) [inline lib/puppet/type/keystore_certificate.rb:61]: Is this correct? Shouldn't the keystore be created before the certificate is added? Otherwise you risk incorrect permissions which can leave the keystore world readable. Probably not a huge deal since we still use password files but still.

File: manifests/candlepin.pp
Comment (ekohl) [inline manifests/candlepin.pp:107]: Did we not manage these before? Did that leave the key world readable?

Comment (ehelms) [reply manifests/candlepin.pp:107]: The `privkey` provider ensures a sane default if we do not manage the key through a file resource:

https://github.com/theforeman/puppet-certs/blob/master/lib/puppet/provider/privkey/katello_ssl_tool.rb#L33

```
# ls -la /etc/pki/katello/private/katello-tomcat.key 
-r--------. 1 root root 3243 May 21 13:53 /etc/pki/katello/private/katello-tomcat.key
```

File: lib/puppet/provider/keystore_certificate/openssl.rb
Comment (ehelms) [reply lib/puppet/provider/keystore_certificate/openssl.rb:14]: The latter. We have two existence conditions currently:

1) Does the specified alias exist in the specified keystore
2) Does the specified certificate fingerprint match what is in the keystore

File: lib/puppet/type/keystore_certificate.rb
Comment (ehelms) [reply lib/puppet/type/keystore_certificate.rb]: Yea, this was a leftover from the previous implementation that I didn't question until you raised it. I will drop it.

File: lib/puppet/provider/keystore_certificate/openssl.rb
Comment (wbclark) [inline lib/puppet/provider/keystore_certificate/openssl.rb:61]:
```suggestion
  def add_certificate
    Tempfile.open('temp_keystore') do |temp_store|
      openssl(
        'pkcs12',
        '-export',
        '-in', resource[:certificate],
        '-inkey', resource[:private_key],
        '-out', temp_store,
        '-name', resource[:alias],
        '-CAfile', resource[:ca],
        '-password', "file:#{resource[:password_file]}"
      )
      rescue Puppet::ExecutionFailure => e
        Puppet.err("Failed to export certificate to PKCS12: #{e}")
        return
      end

      keytool(
        '-importkeystore',
        '-noprompt',
        '-srckeystore', temp_store,
        '-srcstorepass:file', resource[:password_file],
        '-destkeystore', resource[:keystore],
        '-deststorepass:file', resource[:password_file],
        '-srcalias', resource[:alias],
        '-destalias', resource[:alias]
      )
      rescue Puppet::ExecutionFailure => e
        Puppet.err("Failed to add certificate to keystore: #{e}")
        return
      end
    end
  end
```

This changes:

1. `def add_certificate` was missing `end`

2. Log issues here at error level

3. Move `rescue` blocks into method, and `return`

(I haven't yet tested this)

Comment (wbclark) [reply lib/puppet/provider/keystore_certificate/openssl.rb:61]: I don't this is quite right. I see now the intended structure was

```
def add_certificate
  Tempfile.open('temp_keystore') do |temp_store|
    openssl(...)
    keytool(...)
  rescue
    ...
  end
end
```

Comment (ehelms) [reply lib/puppet/provider/keystore_certificate/openssl.rb:61]: I can make the change to error level. The two operations are "atomic" in their execution given the Temp file that is being created. The code gets kinda of ugly attempting to rescue multiple places and ensure the Temp file is cleaned up in each and every case. The failure being caught for either invocation would trigger the rescue and printing useful information to the user. That is why I chose not to split it up.

**Review by ekohl — COMMENTED** [review]
Small nits, otherwise :+1:

File: manifests/candlepin.pp
Comment (ekohl) [reply manifests/candlepin.pp:100]: Looks like you modified one that I didn't spot, but not this particular instance ;)

File: lib/puppet/provider/keystore_certificate/openssl.rb
Comment (wbclark) [inline lib/puppet/provider/keystore_certificate/openssl.rb:89]: After a bit of testing, I understand that the second `&.first` is necessary because otherwise idempotence is broken:

```
Notice: /Stage[main]/Certs::Candlepin/Keystore_certificate[/etc/candlepin/certs/keystore:tomcat]/certificate: certificate changed ['FD:64:E2:53:2D:2D:30:AC:8D:C7:9A:D4:9C:D8:DD:D1:3D:09:98:B5:41:9D:F0:ED:66:9F:40:C2:7B:13:7C:7E'] to 'FD:64:E2:53:2D:2D:30:AC:8D:C7:9A:D4:9C:D8:DD:D1:3D:09:98:B5:41:9D:F0:ED:66:9F:40:C2:7B:13:7C:7E'
```

I had a concern about multiple certificates in the keystore, but we are specifying the `-alias` so we should get only the one we care about

File: manifests/candlepin.pp
Comment (wbclark) [reply manifests/candlepin.pp:140]: I tested this, but:

```
  Error: Evaluation Error: Error while evaluating a Resource Statement, No set of title patterns matched the title "funny_name". (file: /etc/puppetlabs/code/environments/production/modules/certs/manifests/candlepin.pp, line: 140, column: 5) on node centos7-64-1
```

My code was:

```puppet
    keystore_certificate { "funny_name":
      keystore      => $keystore,
      alias         => 'tomcat',
      ensure        => present,
      password_file => $keystore_password_path,
      certificate   => $tomcat_cert,
      private_key   => $tomcat_key,
      ca            => $ca_cert,
    }
```

File: lib/puppet/provider/keystore_certificate/openssl.rb
Comment (ehelms) [reply lib/puppet/provider/keystore_certificate/openssl.rb:89]: That's the idea.

File: manifests/candlepin.pp
Comment (ehelms) [reply manifests/candlepin.pp]: I've updated this to support passing:

  1) either just `keystore` in which case the name is the `alias`
  2) `keystore:alias`
  3) `keystore` and `alias` parameters with a random name

My workflow is adding some stand alone provider tests in follow up PRs so I will handle testing some of that in there.

**Comment by ehelms** [comment]
@wbclark @evgeni Been a wee bit, y'all reviewed this originally can you come back to it?

File: lib/puppet/type/keystore_certificate.rb
Comment (wbclark) [inline lib/puppet/type/keystore_certificate.rb:9]: Do we expect the possibility of multi-line resource titles?

Further  below (L13) do do we intend to map multi-line resource titles to an alias only?

Some elaboration on the choice of `/.../m` would be helpful to review this

Comment (ehelms) [reply lib/puppet/type/keystore_certificate.rb:9]: I should know this but what is a multi-line resource title?

Comment (ehelms) [reply lib/puppet/type/keystore_certificate.rb:9]: I did it this way to support `/path/to/keystore:alias` and `alias` as titles.

Comment (wbclark) [reply lib/puppet/type/keystore_certificate.rb:9]: I'm reading at https://ruby-doc.org/core-2.4.1/Regexp.html#EXTENDED

"/pat/m - Treat a newline as a character matched by ."

How does it relate to `/path/to/keystore:alias` ? There is some problem with the `/` characters in the path?

Comment (ehelms) [reply lib/puppet/type/keystore_certificate.rb:9]: I didn't much think about it. I copied this pattern from the `truststore_certificate` we had built previously, and that originates from https://github.com/theforeman/puppet-certs/pull/312

File: manifests/candlepin.pp
Comment (ekohl) [reply manifests/candlepin.pp:126]: Marking as unresolved. While that PR was merged, this now drops the `show_diff` again.

File: manifests/candlepin.pp
Comment (ekohl) [reply manifests/candlepin.pp:140]: Can we mark this as resolved now?

**Review by wbclark — COMMENTED** [review]
We have both a `$keystore_password_file` and `$keystore_password_path` in manifests/candlepin.pp

The `$keystore_password_file` is only used in `$keystore_password = extlib::cache_data('foreman_cache_data', $keystore_password_file, extlib::random_password(32))`

Is that intended as a transitionary design to move the location while keeping the same content, or is it simply an oversight?

**Comment by ehelms** [comment]
> We have both a `$keystore_password_file` and `$keystore_password_path` in manifests/candlepin.pp
> 
> The `$keystore_password_file` is only used in `$keystore_password = extlib::cache_data('foreman_cache_data', $keystore_password_file, extlib::random_password(32))`
> 
> Is that intended as a transitionary design to move the location while keeping the same content, or is it simply an oversight?

When using `cache_data` you are almost doing a key-value scenario where you give it a filename (key) and the password (value) you want to store by that name in the caching location. So you end up with something like:

```
/opt/puppetlabs/puppet/cache/foreman_cache_data/keystore_password-file
```

If you were instead to pass it `$keystore_password_path`, you would end up with:

```
/opt/puppetlabs/puppet/cache/foreman_cache_data/etc/pki/katello/keystore_password-file
```

**Comment by evgeni** [comment]
Doesn't that mean we can't have two keystores at `/etc/pki/katello/keystore` and `/etc/pki/candlepin/keystore` (stupid example, but two stores with the same `basemame`)?

**Comment by ehelms** [comment]
> Doesn't that mean we can't have two keystores at `/etc/pki/katello/keystore` and `/etc/pki/candlepin/keystore` (stupid example, but two stores with the same `basemame`)?

Nothing prevents this. If this isn reference to the `cache_data`, you just have to ensure the cache key is different for each keystores password or that you change the namespace. For example, if `/etc/pki/candlepin/keystore` was instead defined in `puppet-candlepin` you might should opt to choose `candlepin_cache_data` as your namespace.

**Comment by wbclark** [comment]
Got it, thanks.

Instead of:

```puppet
  $keystore_password = extlib::cache_data('foreman_cache_data', $keystore_password_file, extlib::random_password(32))
  $truststore_password = extlib::cache_data('foreman_cache_data', $truststore_password_file, extlib::random_password(32))    
  $keystore_password_path = "${pki_dir}/keystore_password-file"
  $truststore_password_path = "${pki_dir/truststore_password-file"
```

I believe this would be a cleaner alternative:

```puppet
  $keystore_password = extlib::cache_data('foreman_cache_data', $keystore_password_file, extlib::random_password(32))
  $truststore_password = extlib::cache_data('foreman_cache_data', $truststore_password_file, extlib::random_password(32))    
  $keystore_password_path = "${pki_dir}/${keystore_password_file}"
  $truststore_password_path = "${pki_dir/${truststore_password_file}"
```

So at least the path variable would be built from the file variable.

**Comment by ehelms** [comment]
Oh sure I can close that previous gap in how things were defined. Updated.

**Comment by ehelms** [comment]
I believe I have now addressed everything. Could I get another review?

**Review by ekohl — APPROVED** [review]
I'd also consider checking file paths in the type are absolute paths, but I won't block on that.

**Review by wbclark — APPROVED** [review]
I'd like to better understand the implementation choices in the `title_patterns` but since it was copied from an already working type and tests validate that it's working here as well, I think I can treat that as my own homework -- starting with adding test cases for the various title patterns. I don't see that as something that needs to block this work, as the primary case that we care about is already validated.

**Pull request merged**
